File: app/src/main/kotlin/io/builderscon/conference2017/viewmodel/SessionViewModel.java
package io.builderscon.conference2017.viewmodel;

import android.content.Context;
import android.content.Intent;
import android.databinding.BaseObservable;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.format.DateFormat;
import android.view.View;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import io.builderscon.client.model.Session;
import io.builderscon.conference2017.R;
import io.builderscon.conference2017.view.activity.SessionDetailActivity;

// TODO convert to kotlin
public class SessionViewModel extends BaseObservable {

    private Session session;

    private String shortStime = "";

    private String title = "";

    private String roomName = "";

    private String minutes = "";

    private String avatarURL = "";

    private int rowSpan = 1;

    private int colSpan = 1;

    private int titleMaxLines = 5;

    @DrawableRes
    private int backgroundResId;

    private boolean isClickable = true;

    SessionViewModel(@NonNull Session session, Context context) {
        this.session = session;
        this.shortStime = DateUtil.getHourMinute(session.getStartsOn());
        this.title = session.getTitle();

        this.avatarURL = session.getSpeaker().getAvatarURL();
        this.roomName = session.getRoom().getName();
        this.minutes = "(" + session.getDuration() / 60 + " 分)";

        decideRowSpan(session);
    }

    private SessionViewModel(int rowSpan, int colSpan) {
        this.rowSpan = rowSpan;
        this.colSpan = colSpan;
        this.isClickable = false;
        this.avatarURL = "";
        this.backgroundResId = R.drawable.bg_empty_session;
    }

    static SessionViewModel createEmpty(int rowSpan) {
        return createEmpty(rowSpan, 1);
    }

    static SessionViewModel createEmpty(int rowSpan, int colSpan) {
        return new SessionViewModel(rowSpan, colSpan);
    }

    private void decideRowSpan(@NonNull Session session) {
        if (session.getDuration() / 60 > 30) {
            this.rowSpan = this.rowSpan * 2;
            this.titleMaxLines = this.titleMaxLines * 2;
        }
    }

    Date getStime() {
        return session.getStartsOn();
    }

    public void showSessionDetail(@SuppressWarnings("unused") View view) {
        Intent intent = new Intent(view.getContext(), SessionDetailActivity.class);
        intent.putExtra("abstract", session.getAbstract());
        intent.putExtra("avatarURL", session.getSpeaker().getAvatarURL());
        intent.putExtra("speakerName", session.getSpeaker().getNickname());
        intent.putExtra("title", session.getTitle());
        intent.putExtra("start", DateUtil.getLongFormatDate(session.getStartsOn()));
        intent.putExtra("minutes", session.getDuration() / 60 + "分");
        intent.putExtra("roomName", roomName);
        intent.putExtra("materialLevel", getLevelTag(session.getMaterialLevel().toString()));
        view.getContext().startActivity(intent);
    }

    // TODO Refactor
    private String getLevelTag(String level) {
        if (level.equals("INTERMEDIATE")) return "中級者";
        if (level.equals("ADVANCED")) return "上級者";
        return "初心者";

    }

    public String getShortStime() {
        return shortStime;
    }

    public String getTitle() {
        return title;
    }

    String getRoomName() {
        return roomName;
    }

    public String getMinutes() {
        return minutes;
    }

    public int getRowSpan() {
        return rowSpan;
    }

    public int getColSpan() {
        return colSpan;
    }

    public int getBackgroundResId() {
        return backgroundResId;
    }

    public boolean isClickable() {
        return isClickable;
    }

    public int getTitleMaxLines() {
        return titleMaxLines;
    }

    public String getAvatarURL() {
        return avatarURL;
    }

}

class DateUtil {

    private static final String FORMAT_KKMM = "kk:mm";

    private static final String FORMAT_YYYYMMDDKKMM = "yyyyMMMdkkmm";

    private DateUtil() {
        throw new AssertionError("no instance!");
    }

    @NonNull
    public static String getHourMinute(Date date) {
        String pattern = DateFormat.getBestDateTimePattern(Locale.getDefault(), FORMAT_KKMM);
        return new SimpleDateFormat(pattern, Locale.getDefault()).format(date);

    }

    @NonNull
    public static String getLongFormatDate(@Nullable Date date) {
        if (date == null) {
            return "";
        }

        String pattern = DateFormat.getBestDateTimePattern(Locale.getDefault(), FORMAT_YYYYMMDDKKMM);
        return new SimpleDateFormat(pattern, Locale.getDefault()).format(date);

    }
}

File: app/src/main/kotlin/io/builderscon/conference2017/viewmodel/SessionViewModel.kt
package io.builderscon.conference2017.viewmodel

import android.content.Intent
import android.databinding.BaseObservable
import android.support.annotation.DrawableRes
import android.view.View
import io.builderscon.client.model.Session
import io.builderscon.conference2017.R
import io.builderscon.conference2017.extension.getHourMinute
import io.builderscon.conference2017.extension.getLongFormatDate
import io.builderscon.conference2017.extension.needsAdjustColSpan
import io.builderscon.conference2017.view.activity.SessionDetailActivity
import java.util.*

class SessionViewModel(private val session: Session?) : BaseObservable() {

    private var shortStime = ""

    private var title = ""

    internal var roomName = ""

    private var minutes = ""

    private var avatarURL = ""

    internal var rowSpan = 1

    private var colSpan = 1

    private var titleMaxLines = 5

    @DrawableRes
    private var backgroundResId: Int = 0

    private var isClickable = true

    init {
        session?.let {
            this.shortStime = it.startsOn.getHourMinute()
            this.title = it.title
            this.avatarURL = it.speaker.avatarURL
            this.roomName = it.room.name
            this.minutes = "(" + it.duration / 60 + " 分)"
            if (it.startsOn.needsAdjustColSpan()) this.colSpan = 2

            decideRowSpan(it)
        }

    }

    private fun decideRowSpan(session: Session) {
        if (session.duration / 60 > 30) {
            this.rowSpan = this.rowSpan * 2
            this.titleMaxLines = this.titleMaxLines * 2
        }
    }

    fun getStime(): Date? {
        return session?.startsOn
    }

    fun showSessionDetail(view: View) {
        val intent = Intent(view.context, SessionDetailActivity::class.java)
        session?.let {
            intent.putExtra("abstract", session.abstract)
            intent.putExtra("avatarURL", session.speaker.avatarURL)
            intent.putExtra("speakerName", session.speaker.nickname)
            intent.putExtra("title", session.title)
            intent.putExtra("start", session.startsOn.getLongFormatDate())
            intent.putExtra("minutes", (session.duration / 60).toString() + "分")
            intent.putExtra("roomName", roomName)
            intent.putExtra("materialLevel", getLevelTag(session.materialLevel.toString()))
        }
        view.context.startActivity(intent)
    }

    // TODO Refactor
    private fun getLevelTag(level: String): String {
        if (level == "INTERMEDIATE") return "中級者"
        if (level == "ADVANCED") return "上級者"
        return "初級者"
    }

    fun getShortStime(): String {
        return shortStime
    }

    fun getTitle(): String {
        return title
    }

    fun getMinutes(): String {
        return minutes
    }

    fun getRowSpan(): Int {
        return rowSpan
    }

    fun getColSpan(): Int {
        return colSpan
    }

    fun getBackgroundResId(): Int {
        return backgroundResId
    }

    fun isClickable(): Boolean {
        return isClickable
    }

    fun getTitleMaxLines(): Int {
        return titleMaxLines
    }

    fun getAvatarURL(): String {
        return avatarURL
    }

    companion object SessionViewModel {
        fun createEmpty(rowSpan: Int): io.builderscon.conference2017.viewmodel.SessionViewModel {
            return createEmpty(rowSpan, 1)
        }

        fun createEmpty(rowSpan: Int, colSpan: Int): io.builderscon.conference2017.viewmodel.SessionViewModel {
            val empty = SessionViewModel(null)
            return empty.apply {
                empty.rowSpan = rowSpan
                empty.colSpan = colSpan
                isClickable = false
                avatarURL = ""
                backgroundResId = R.drawable.bg_empty_session
            }
        }
    }

}

